Add search queries to PBC interface.

seancribbs opened the pull request:
This is part of a series of pull-requests that adds Riak Search features to the Protocol Buffers interface. It requires the existing refactoring pull-requests to be merged first.

Related: basho/riak_pb#2, [basho/riak_api](/basho/riak_api)
- We are not adding indexing or deleting at this time because we want to encourage use of the KV integration instead.
- All options of the "solr" interface are supported, although "q.op" is named "op". "fl" is expressed as a repeated fields instead of as a comma-separated list.
- Unlike the "solr" interface, the additional search parameter information is not returned, which mostly just repeats request information. Only MaxScore and NumFound are returned to the client.

The gist below contains a patch against basho/riak-ruby-client to demo the new feature and a preliminary, unscientific benchmark script and result set. It relies on loading the "spam" corpus from `basho_expect`. Results suggest (aside from inefficiencies on the Ruby side) that the overhead of HTTP dominates. I will update shortly with comparisons to the MapReduce emulation.

https://gist.github.com/cb87a5c0447e91af53da

Commit:
Add search queries to PBC interface.

* We are not adding indexing or deleting at this time because we want
  to encourage use of the KV integration instead.
* All options of the "solr" interface are supported, although "q.op"
  is named "op".
* Unlike the "solr" interface, the additional search parameter
  information is not returned, which mostly just repeats request
  information. Only MaxScore and NumFound are returned to the client.

rzezeski commented on rebar.config:
```diff
@@ -13,7 +13,9 @@
-                                {branch, "master"}}}
+                                {branch, "master"}}},
+       {riak_api, ".*", {git, "git://github.com/basho/riak_api", {branch, "master"}}},
+       {riak_pb, ".*", {git, "git://github.com/basho/riak_pb", {branch, "sdc-pb-search"}}}
```
This should be master branch.

seancribbs replied:
Ok, but you'll have to manually check out the related branch.

Commit:
Use master branch for riak_pb.

seancribbs commented:
In general there are some things in here that could be extracted from the WM resource and the PB service into a unified interface. Suggestions welcomed.

rzezeski commented on src/riak_search_pb_query.erl:
```diff
@@ -0,0 +1,140 @@
+-import(riak_pb_search_codec, [encode_search_doc/1]).
+
+-record(?MODULE, {client}).
+-define(state, #?MODULE).
```
This is a non-idiomatic way to declare state.  Why not just stick with the typical `#state{}` record idiom?  I see you have to return it the `pb_service` behavior, that also seems weird.  Why is a record defined by this module being passed back to it's behavior?

Correction, returning the state to behavior isn't weird since it's keeping track of it...I just got confused by this style and figured maybe the behavior is interrogating the state in some way.

rzezeski commented on src/riak_search_pb_query.erl:
```diff
@@ -0,0 +1,140 @@
+                    {error, "Missing query", State}
+            end;
+        Error ->
+            {error, {format, "Could not parse schema '~s': ~p", [Index, Error]}, State}
```
This `Error` won't necessarily always be a parse error, just the most common.

rzezeski commented on src/riak_search_pb_query.erl:
```diff
@@ -0,0 +1,140 @@
+         process/2,
+         process_stream/3]).
+
+-import(riak_search_utils, [to_atom/1, to_integer/1, to_binary/1, to_boolean/1, to_float/1]).
```
`to_integer` and `to_boolean` are not being used.  I also tend not to use imports, but that's just personal preference.

Commit:
Use more idiomatic state record instead of module name and macro.

Commit:
to_integer/1 and to_boolean/1 are unused.

Commit:
Schema error is not always a parsing one.

Commit:
Add license header and docstring.

rzezeski committed:
Remove trailing whitespace

rzezeski committed:
Move common HTTP/PB `run_query`

Centralize the common logic between HTTP & PB query execution.

rzezeski committed:
Use common fun for PB query

Use the common `run_query` for the PB query execution.

rzezeski committed:
Fix dialyzer errors

* The `search_field` can be binary because of inline fields code.

* Typo in `riak_search_test`

rzezeski commented on src/riak_search_pb_query.erl:
```diff
@@ -0,0 +1,153 @@
+                {error, missing_query} ->
+                    {error, "Missing query", State}
+            end;
+        Error ->
```
Actually, on closer look `get_schema` is either going to return `{ok, Schema}` or throw.  So either turn this into a catch or just take the happy path.  Personally I would vote happy path for now as `get_schema` should almost always return.

seancribbs replied:
So should we `catch` the expression inside the case header?

rzezeski replied:
I'm saying forget the case statement and just do `{ok, Schema} = ...`  If that call fails something really stupid happened (well actually, the ETS table for the cache may not be around if these calls are somehow coming in before Search starts but that should be a change to get_schema, not here).

seancribbs replied:
They shouldn't come around before search starts, but I haven't 100% verified that yet (just basing it on what others have said about reltool). Either way, if search waits properly for its vnodes to start like KV does, the PB service won't be registered until it's up.

The other aspect is that I wanted to avoid dumping the connection except in situations where the service state is completely broken, so the callbacks should try to return `error` tuples instead of tossing exceptions. Uncaught errors go through a try/catch clause that warns the client of the error, logs it at the debug level and then stops the TCP server.

rzezeski replied:
In this case the state is severely broken if get_schema throws and I think the right thing to do is drop the conn.  It seems like defensive programming for no benefit to me.  In any case, this current code is wrong.  Either it needs to catch or do `{ok, Schema}` match only.

seancribbs replied:
Noted, will fix.

rzezeski committed:
Move `replace_schema_defaults` into utils

This is another function that is shared between PB & HTTP.

rzezeski committed:
Remove defensive programming

1. There is no chance for `get_schema` to return an error tuple when
passed an index.  Instead it will throw.

2. It will only throw if something is horribly wrong, so let it crash
the process and kill the conn rather than doing defensive programming.

rzezeski committed:
Fix `encode_results` and FL

* The `encode_search_doc` expects a proplist, not a tuple.

* Need to parse FL.

rzezeski commented:
+1 to merge this

seancribbs commented:
Waiting to merge until other refactorings land.
